packages/x-date-pickers/src/DateCalendar/DateCalendar.tsx

```diff
@@ -73,6 +74,7 @@ const DateCalendarRoot = styled(PickerViewRoot, {
 })<{ ownerState: DateCalendarProps<any> }>({
   display: 'flex',
   flexDirection: 'column',
+  height: VIEW_HEIGHT,
```

Comment:
This is the main change that fixes the underlying problem.
Everything else is just tying things together.

Reply:
I'm slightly worried about the `MonthCalendar` in this scenario (or the `YearCalendar` if there aren't too many options) because of the white space. 
![image](https://github.com/mui/mui-x/assets/72460825/4cf7f0b1-0726-46ff-bdde-30d784e66d38)

But I think it's a good compromise considering the number of users who would actually  end up in this scenario vs the problem this change solves 👍

Reply:
I understand the concern, but in our case, we do not even use the `monthsPerRow = 4` in case of Desktop pickers, which we should probably do, to keep in line with spacing and sizing convention based on MD2.
WDYT? 🤔 
Does seeing the same approach on https://material.angular.io/components/datepicker/overview#setting-the-calendar-starting-view alleviate some of the concerns? 🤔

Reply:
Having seen another example for a similar approach is definitely reassuring 👍 
I think this solution is definitely solving a lot of issues, so my concern is mostly something I wanted to highlight, and definitely shouldn't be a blocker 💯